File: craps/events.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Callable, Dict, List, Optional, Tuple, Type, Union


@dataclass(frozen=True)
class Event:
    """Base class for all engine events."""


@dataclass(frozen=True)
class SessionStarted(Event):
    num_shooters: int


@dataclass(frozen=True)
class ShooterAssigned(Event):
    shooter_index: int
    shooter_name: str


@dataclass(frozen=True)
class BetsRequested(Event):
    """The engine is about to collect bets from strategies."""


@dataclass(frozen=True)
class BetPlaced(Event):
    player_name: str
    bet_type: str
    amount: int
    number: Optional[Union[int, Tuple[int, int]]]


@dataclass(frozen=True)
class DiceRolled(Event):
    shooter_index: int
    roll_number: int
    dice: Tuple[int, int]
    total: int
    phase: str
    point: Optional[int]
    table_risk: int = 0
    shooter_name: str = ""


@dataclass(frozen=True)
class BetResolved(Event):
    player_name: str
    bet_type: str
    amount: int
    number: Optional[Union[int, Tuple[int, int]]]
    status: str
    payout: int
    #: Bet.payout() at resolution time — what the stats ledger credits on a
    #: win (may include the returned stake for contract bets).
    win_payout: int = 0
    #: Whether settlement took the bet off the table. Winning non-contract
    #: bets can stay up ("if it pays, it stays" per house rules); the felt
    #: must not have to know that rule — the engine states the fact.
    removed: bool = False


@dataclass(frozen=True)
class BetMoved(Event):
    """A bet acquired its number after placement: Come/Don't Come travel,
    or an odds bet attaching to the established point."""
    player_name: str
    bet_type: str
    amount: int
    number: int


@dataclass(frozen=True)
class BetAdjusted(Event):
    """A live bet was reshaped between rolls (pressing/regression or an
    on/off flip requested by the strategy)."""
    player_name: str
    bet_type: str
    amount: int
    number: Optional[Union[int, Tuple[int, int]]]
    status: str


@dataclass(frozen=True)
class BetStatusChanged(Event):
    """A live bet's working status changed outside resolution (phase
    refresh or a strategy status spec at bet-collection time)."""
    player_name: str
    bet_type: str
    number: Optional[Union[int, Tuple[int, int]]]
    status: str


@dataclass(frozen=True)
class NumberHit(Event):
    """ATS tracking recorded a number this roll (message pre-formatted)."""
    total: int
    message: str


@dataclass(frozen=True)
class GameStateChanged(Event):
    """Phase/point transition after resolution (message pre-formatted)."""
    message: str


@dataclass(frozen=True)
class BankrollsUpdated(Event):
    """Per-player bankrolls after post-roll adjustments, in lineup order."""
    bankrolls: Tuple[Tuple[str, int], ...]


@dataclass(frozen=True)
class RiskUpdated(Event):
    """Per-player active amounts at risk after status refresh, in lineup order."""
    at_risk: Tuple[Tuple[str, int], ...]


@dataclass(frozen=True)
class PointEstablished(Event):
    point: int


@dataclass(frozen=True)
class PointHit(Event):
    point: int


@dataclass(frozen=True)
class SevenOut(Event):
    shooter_index: int
    #: Per-player bankroll delta over this shooter's hand, in lineup order.
    shooter_results: Tuple[Tuple[str, int], ...] = ()


@dataclass(frozen=True)
class SessionFinalized(Event):
    session_rolls: int


EventHandler = Callable[[Event], None]


class EventBus:
    """Minimal synchronous pub/sub bus.

    Handlers subscribe by event class and receive every published event of
    that class or any subclass — subscribing to ``Event`` observes the full
    stream. Dispatch is synchronous and in subscription order, keeping
    session runs deterministic.
    """

    def __init__(self) -> None:
        self._handlers: Dict[Type[Event], List[EventHandler]] = {}

    def subscribe(self, event_type: Type[Event], handler: EventHandler) -> None:
        self._handlers.setdefault(event_type, []).append(handler)

    def publish(self, event: Event) -> None:
        for cls in type(event).__mro__:
            if cls is object:
                continue
            for handler in self._handlers.get(cls, ()):
                handler(event)

File: web/src/lib/events.ts
/**
 * The wire vocabulary (Phase 2, D3): TypeScript mirror of the frozen
 * dataclasses in craps/events.py. Field names match the Python event
 * fields exactly — the Python vocabulary IS the wire schema. Every
 * envelope is `{seq, table_id, type, ...payload}`.
 *
 * JSON has no tuples, so tuple-shaped Python fields arrive as arrays:
 * dice pairs, per-player pair lists, and hop-bet numbers like [3, 3].
 */

/** int | hop pair | not-yet-numbered */
export type BetNumber = number | [number, number] | null

interface Base {
  seq: number
  table_id: string
}

export interface SessionStarted extends Base {
  type: 'SessionStarted'
  num_shooters: number
}

export interface ShooterAssigned extends Base {
  type: 'ShooterAssigned'
  shooter_index: number
  shooter_name: string
}

export interface BetsRequested extends Base {
  type: 'BetsRequested'
}

export interface BetPlaced extends Base {
  type: 'BetPlaced'
  player_name: string
  bet_type: string
  amount: number
  number: BetNumber
}

export interface BetMoved extends Base {
  type: 'BetMoved'
  player_name: string
  bet_type: string
  amount: number
  number: number
}

export interface BetAdjusted extends Base {
  type: 'BetAdjusted'
  player_name: string
  bet_type: string
  amount: number
  number: BetNumber
  status: string
}

export interface BetStatusChanged extends Base {
  type: 'BetStatusChanged'
  player_name: string
  bet_type: string
  number: BetNumber
  status: string
}

export interface DiceRolled extends Base {
  type: 'DiceRolled'
  shooter_index: number
  roll_number: number
  dice: [number, number]
  total: number
  phase: 'come-out' | 'point'
  point: number | null
  table_risk: number
  shooter_name: string
}

export interface BetResolved extends Base {
  type: 'BetResolved'
  player_name: string
  bet_type: string
  amount: number
  number: BetNumber
  status: string
  payout: number
  win_payout: number
  /** Whether settlement took the chip off the table ("if it pays, it
   * stays" winners keep their felt position). */
  removed: boolean
}

export interface NumberHit extends Base {
  type: 'NumberHit'
  total: number
  message: string
}

export interface GameStateChanged extends Base {
  type: 'GameStateChanged'
  message: string
}

export interface BankrollsUpdated extends Base {
  type: 'BankrollsUpdated'
  bankrolls: [string, number][]
}

export interface RiskUpdated extends Base {
  type: 'RiskUpdated'
  at_risk: [string, number][]
}

export interface PointEstablished extends Base {
  type: 'PointEstablished'
  point: number
}

export interface PointHit extends Base {
  type: 'PointHit'
  point: number
}

export interface SevenOut extends Base {
  type: 'SevenOut'
  shooter_index: number
  shooter_results: [string, number][]
}

export interface SessionFinalized extends Base {
  type: 'SessionFinalized'
  session_rolls: number
}

export type Envelope =
  | SessionStarted
  | ShooterAssigned
  | BetsRequested
  | BetPlaced
  | BetMoved
  | BetAdjusted
  | BetStatusChanged
  | DiceRolled
  | BetResolved
  | NumberHit
  | GameStateChanged
  | BankrollsUpdated
  | RiskUpdated
  | PointEstablished
  | PointHit
  | SevenOut
  | SessionFinalized

export const EVENT_TYPES = [
  'SessionStarted',
  'ShooterAssigned',
  'BetsRequested',
  'BetPlaced',
  'BetMoved',
  'BetAdjusted',
  'BetStatusChanged',
  'DiceRolled',
  'BetResolved',
  'NumberHit',
  'GameStateChanged',
  'BankrollsUpdated',
  'RiskUpdated',
  'PointEstablished',
  'PointHit',
  'SevenOut',
  'SessionFinalized',
] as const satisfies readonly Envelope['type'][]
